### PR TITLE
Add DumpDelegate sos command

### PR DESCRIPTION
**Pull request by stephentoub**
Finding out what method a delegate points to today with SOS is tedious.  This PR adds a DumpDelegate command that simplifies it: give it the delegate address, and it outputs a line for each delegate that makes up the input delegate, where each line includes the target object, the method descriptor, and a friendly name.

For example, running this C# program and attaching WinDBG:
```C#
using System;
using System.Diagnostics;

class Program
{
    static void Main()
    {
        Action sm = StaticMethod;
        Action im = new Program().InstanceMethod;
        Action combined = sm + im + sm + im;
        Console.ReadLine();
    }

    static void StaticMethod() { }
    void InstanceMethod() { }
}
```
yields:
image

cc: @janvorli, @mikem8361, @noahfalk, @jkotas

### CODE REVIEW

File: src/ToolBox/SOS/Strike/util.cpp [hunk 1]
```diff
@@ -2468,6 +2468,77 @@ BOOL IsStringObject (size_t obj)
...
+    int offset;
+
+    if ((offset = GetObjFieldOffset(delegateObj.GetAddress(), delegateObj.GetMT(), W("_methodPtrAux"))) != 0)
+    {
+        CLRDATA_ADDRESS methodPtrAux;
+        MOVE(methodPtrAux, delegateObj.GetAddress() + offset);
+        if (methodPtrAux != NULL)
+        {
+            DacpCodeHeaderData codeHeaderData;
+            if (codeHeaderData.Request(g_sos, methodPtrAux) == S_OK)
```
Comment (jkotas) [inline src/ToolBox/SOS/Strike/util.cpp]: This should rather use `GetMethodDescPtrFromIP` like below. The address does not have to point to code header.

File: src/ToolBox/SOS/Strike/util.cpp [hunk 2]
```diff
@@ -2468,6 +2468,77 @@ BOOL IsStringObject (size_t obj)
...
+        CLRDATA_ADDRESS methodPtr;
+        MOVE(methodPtr, delegateObj.GetAddress() + offset);
+        if (methodPtr != NULL)
+        {
+            if (g_sos->GetMethodDescPtrFromIP(methodPtr, pMD) == S_OK)
+            {
+                return TRUE;
+            }
+
+            DacpCodeHeaderData codeHeaderData;
```
Comment (jkotas) [inline src/ToolBox/SOS/Strike/util.cpp]: This should not be necessary. `GetMethodDescPtrFromIP` should be superset of what this does.

File: src/ToolBox/SOS/Strike/util.cpp (hunk 1 above)
Comment (stephentoub) [reply src/ToolBox/SOS/Strike/util.cpp]: Thanks.  Changed.

File: src/ToolBox/SOS/Strike/util.cpp (hunk 2 above)
Comment (stephentoub) [reply src/ToolBox/SOS/Strike/util.cpp]: Ok.  Deleted.

File: src/ToolBox/SOS/Strike/util.cpp [hunk 3]
```diff
@@ -2498,41 +2498,26 @@ BOOL TryGetMethodDescriptorForDelegate(CLRDATA_ADDRESS delegateAddr, CLRDATA_ADD
...
         MOVE(methodPtr, delegateObj.GetAddress() + offset);
-        if (methodPtr != NULL)
+        if (methodPtr != NULL && g_sos->GetMethodDescPtrFromIP(methodPtr, pMD) == S_OK)
         {
-            if (g_sos->GetMethodDescPtrFromIP(methodPtr, pMD) == S_OK)
-            {
-                return TRUE;
-            }
-
-            DacpCodeHeaderData codeHeaderData;
```
Comment (stephentoub) [inline src/ToolBox/SOS/Strike/util.cpp]: @jkotas, I removed these CodeHeaderData check based on your feedback, but I'm seeing common cases where without it I'm unable to resolve the delegate and with it I am; in those cases ip2md similarly fails.  It sounds like that's not expected?


As an example, I've got a simple ASP.NET Core Web App.  I just created a new one with the template, added this for the index page:
```C#
using System.Threading;
using System.Threading.Tasks;
using Microsoft.AspNetCore.Mvc.RazorPages;

namespace WebApplication1.Pages
{
    public class IndexModel : PageModel
    {
        public async Task OnGetAsync()
        {
            await Task.WhenAll(
                Task.Run(() => SomeMethod()),
                Task.Delay(int.MaxValue));
        }

        public static void SomeMethod() => Thread.Sleep(int.MaxValue);
    }
}
```
ran it, and then attached windbg / sos.  Without this additional code, these checks fail.   With them, they successfully resolve "WebApplication1.Pages.IndexModel+<>c.<OnGetAsync>b__0_0()".

Comment (jkotas) [reply src/ToolBox/SOS/Strike/util.cpp]: Yes, you are right. I see that you have added them in your latest PR already. I got confused by the  DAC API names. The names do not match what one would expect the API to do.

**Pull request merged**